src/client/keycode.cpp

```diff
@@ -236,122 +238,122 @@ static const struct table_key table[] = {
+	return invalid_key;
 }
 
-KeyPress::KeyPress(const char *name)
+static const table_key &lookup_scancode(const u32 scancode)
 {
-	if (strlen(name) == 0) {
-		Key = irr::KEY_KEY_CODES_COUNT;
-		Char = L'\0';
-		m_name = "";
-		return;
-	}
+	auto key = RenderingEngine::get_raw_device()->getKeyFromScancode(scancode);
+	return key.index() == 0 ? lookup_keykey(std::get<irr::EKEY_CODE>(key)) : lookup_keychar(std::get<wchar_t>(key));
+}
 
-	if (strlen(name) <= 4) {
-		// Lookup by resulting character
-		int chars_read = mbtowc(&Char, name, 1);
-		FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-		try {
-			auto &k = lookup_keychar(Char);
-			m_name = k.Name;
-			Key = k.Key;
-			return;
-		} catch (UnknownKeycode &e) {};
-	} else {
-		// Lookup by name
-		m_name = name;
-		try {
-			auto &k = lookup_keyname(name);
-			Key = k.Key;
-			Char = k.Char;
-			return;
-		} catch (UnknownKeycode &e) {};
-	}
+KeyPress::KeyPress(std::string_view name)
+{
+	if (loadFromScancode(name))
+		return;
+	const auto &key = lookup_keyname(name);
+	KeyCode keycode(key.Key, key.Char);
+	scancode = RenderingEngine::get_raw_device()->getScancodeFromKey(keycode);
+}
 
-	// It's not a known key, complain and try to do something
-	Key = irr::KEY_KEY_CODES_COUNT;
-	int chars_read = mbtowc(&Char, name, 1);
-	FATAL_ERROR_IF(chars_read != 1, "Unexpected multibyte character");
-	m_name = "";
-	warningstream << "KeyPress: Unknown key '" << name
-		<< "', falling back to first char." << std::endl;
+std::string KeyPress::sym(const bool force_scancode) const
+{
+	if (scancode == 0)
+		return "";
+	if (force_scancode)
+		return formatScancode();
+	const auto &name = lookup_scancode(scancode).Name;
+	if (!name.empty())
+		return name;
+	return formatScancode();
 }
 
-KeyPress::KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character)
+std::string KeyPress::name() const
 {
-	if (prefer_character)
-		Key = irr::KEY_KEY_CODES_COUNT;
-	else
-		Key = in.Key;
-	Char = in.Char;
-
-	try {
-		if (valid_kcode(Key))
-			m_name = lookup_keykey(Key).Name;
-		else
-			m_name = lookup_keychar(Char).Name;
-	} catch (UnknownKeycode &e) {
-		m_name.clear();
-	};
+	const auto &name = lookup_scancode(scancode).LangName;
+	auto table_key = lookup_scancode(scancode);
+	if (!name.empty() || scancode == 0)
+		return name;
+	return formatScancode();
 }
 
-const char *KeyPress::sym() const
+irr::EKEY_CODE KeyPress::getKeycode() const
 {
-	return m_name.c_str();
+	return lookup_scancode(scancode).Key;
 }
 
-const char *KeyPress::name() const
+wchar_t KeyPress::getKeychar() const
 {
-	if (m_name.empty())
-		return "";
-	const char *ret;
-	if (valid_kcode(Key))
-		ret = lookup_keykey(Key).LangName;
-	else
-		ret = lookup_keychar(Char).LangName;
-	return ret ? ret : "<Unnamed key>";
+	return lookup_scancode(scancode).Char;
 }
 
-const KeyPress EscapeKey("KEY_ESCAPE");
+bool KeyPress::loadFromScancode(std::string_view name)
+{
+	if (name.size() < 2 || name[0] != '<')
```

Comment:
> That might be a tad bit cleaner with `std::stoul`.

Not really. I made this stricter by also checking that the end pointer points exactly at the ending `>` character. With `std::stoul` this would not work well because it would additionally involve constructing (and, with said pointer check, keeping) a `std::string`, which could be avoided with `std::strtoul`.